src/Functions/FunctionsStringDistance.cpp

```diff
@@ -271,6 +272,222 @@ struct ByteEditDistanceImpl
     }
 };
 
+struct ByteDamerauLevenshteinDistanceImpl
+{
+    using ResultType = UInt64;
+    static constexpr size_t max_string_size = 1u << 16;
+
+    static ResultType inline process(
+        const char * __restrict haystack, size_t haystack_size, const char * __restrict needle, size_t needle_size)
+    {
+        /// Safety threshold against DoS, since we use two array to calculate the distance.
+        if (haystack_size > max_string_size || needle_size > max_string_size)
+            throw Exception(
+                ErrorCodes::TOO_LARGE_STRING_SIZE,
+                "The string size is too big for function editDistance, "
+                "should be at most {}", max_string_size);
+
+        // short cut cases:
+  	// - null strings
+  	// - zero length strings
+  	// - identical length and value strings
+  	if (haystack == nullptr || needle == nullptr ) return -1;
+  	if (haystack_size == 0) return needle_size;
+  	if (needle_size == 0) return haystack_size;
+  	if (haystack_size == needle_size && memcmp(haystack, needle, haystack_size) == 0) return 0;
+
+  	int i;
+  	int j;
+  	int l_cost;
+  	int ptr_array_length = static_cast<int>(sizeof(int*) * (haystack_size + 1));
+  	int int_array_length = static_cast<int>(sizeof(int) * (needle_size + 1) * (haystack_size + 1));
+	int s1len = static_cast<int>(haystack_size);
+	int s2len = static_cast<int>(needle_size);
+	// Dynamically allocate memory for the 2D array
+  	// Allocating a 2D array (with d being an array of pointers to the start of the rows)
+  	int** d = reinterpret_cast<int**>(new int*[ptr_array_length]);
+  	int* rows = reinterpret_cast<int*>(new int[int_array_length]);
+  	// Setting the pointers in the pointer-array to the start of (s2len + 1) length
+  	// intervals and initializing its values based on the mentioned algorithm.
+  	
+	for (i = 0; i <= s1len; ++i) {
+    		d[i] = rows + (s2len + 1) * i;
+    		d[i][0] = i;
+ 	}
+  	std::iota(d[0], d[0] + s2len + 1, 0);
+
+  	for (i = 1; i <= s1len; ++i) {
+    		for (j = 1; j <= s2len; ++j) {
+      			if (haystack[i - 1] == needle[j - 1]) {
+        			l_cost = 0;
+      			} else {
+        			l_cost = 1;
+      			}
+      			d[i][j] = std::min(d[i - 1][j - 1] + l_cost, // substitution
+                        	 std::min(d[i][j - 1] + 1, // insertion
+                                 d[i - 1][j] + 1) // deletion
+      			);
+      			if (i > 1 && j > 1 && haystack[i - 1] == needle[j - 2]
+          			&& haystack[i - 2] == needle[j - 1]) {
+        				d[i][j] = std::min(d[i][j], d[i - 2][j - 2] + l_cost); // transposition
+      			}
+    		}
+  	}
+	
+  	ResultType result = d[s1len][s2len];
+	// Deallocate memory for each row
+	delete[] rows;
```

Comment:
Hi, manual de-allocation of memory is not allowed in application code. Please use shared_ptr/unique_ptr for allocating memory on heap for automatic de-allocation.
More information about this can be found [here](https://clickhouse.com/docs/en/development/style#how-to-write-code) under memory management.